### PR TITLE
handles watch termination

### PR DESCRIPTION
**Pull request by sunnyagg**
- Handles watch termination with registry for graceful shutdown
- Improved test coverage for leaderelection package

### CODE REVIEW

File: pkg/registry/consulwatch.go [hunk 1]
```diff
@@ -2,22 +2,27 @@ package registry
...
+		plan:   plan,
+		client: client,
 	}
 }
 
 // Handler implements the consul watch handler method and invokes the requester handler
-func (cw *ConsulWatchHandler) Handler(index uint64, result interface{}) {
+func (cwh *ConsulWatcher) handler(index uint64, result interface{}) {
 	pairs, ok := result.(api.KVPairs)
 	if !ok {
```
Comment (Abhishekvrshny) [inline pkg/registry/consulwatch.go:27]: log error, or set error somewhere so that it can be logged outside?

File: pkg/registry/consul.go [hunk 2]
```diff
@@ -110,27 +108,20 @@ func (c *ConsulClient) Release(sessionID string, key string, value string) bool
...
+		"prefix": wh.WatchPath,
 	})
 
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	wh := NewConsulWatchHandler(handler)
+	cwh := NewConsulWatcher(wh, plan, c.client)
```
Comment (Abhishekvrshny) [inline pkg/registry/consul.go:122]: can IWatcher interface be injected in this func. Would help with unit testing

File: service/push-consumer/service.go [hunk 3]
```diff
@@ -86,14 +85,39 @@ func (c *Service) Start() error {
...
+		var err error
+		select {
+		case <-gctx.Done():
+			err = gctx.Err()
+		case <-c.stopCh:
+			err = fmt.Errorf("signal received, stopping push-consumer")
+		}
+
+		if watcher != nil {
+			watcher.StopWatch()
```
[the remark this answers is not in the document]
Comment (sunnyagg) [reply service/push-consumer/service.go:117]: This is just a signal, so this signal will cause the actual watch to terminate which is a blocking watch. we are waiting for Actual watch termination

File: pkg/registry/consul.go (hunk 2 above)
Comment (sunnyagg) [reply pkg/registry/consul.go:122]: This func is being used to create watcher, method can be renamed but not sure if interface injection is possible, we will need to create interface in registry

File: pkg/registry/consulwatch.go (hunk 1 above)
Comment (sunnyagg) [reply pkg/registry/consulwatch.go:27]: will discuss

Comment (sunnyagg) [reply pkg/registry/consulwatch.go:27]: will fix all error handling in next PR, will merge this

**Review by Abhishekvrshny — APPROVED** [review]

**Pull request merged**